lib/minimax.ex

```diff
@@ -0,0 +1,38 @@
+defmodule TicTacToe.Minimax do
+  @player1_token Application.get_env(:tic_tac_toe, :player1_token)
+  @player2_token Application.get_env(:tic_tac_toe, :player2_token)
+
+  def minimax(deps, board, current_player, max_player) do
+    with true <- deps.game_status.is_over(deps.board_manager, board) do
+      score_move(deps.game_status.winner(board), max_player, opponent(max_player))
+    else false ->
+      scored_moves = deps.board_manager.available_positions(board)
+        |> Enum.map(fn position -> 
+        next_board = deps.board_manager.update(board, position, current_player)
+        next_player = opponent(current_player)
+        %{position: position, score: minimax(deps, next_board, next_player, max_player).score}
+      end)
+
+      best_move(scored_moves, max_player, opponent(max_player), current_player)
+    end 
+  end
+
+  defp score_move(winner, max_player, min_player) do
+    case winner do
+      ^max_player -> %{score: 10}
+      ^min_player -> %{score: -10}
+      nil -> %{score: 0}
+    end    
+  end
+
+  defp best_move(scored_moves, max_player, min_player, current_player) do
+    case current_player do
+      ^max_player -> Enum.reduce(scored_moves, %{position: nil, score: -10}, fn (move, acc) -> if move.score >= acc.score, do: move, else: acc end)
+      ^min_player -> Enum.reduce(scored_moves, %{position: nil, score: 10}, fn (move, acc) -> if move.score <= acc.score, do: move, else: acc end)
+    end
+  end
+
+  defp opponent(token) do
```

Comment:
Couldn't you use pattern matching for the args instead of using if statement inside it?

Reply:
I'm not sure how I can achieve the result of the >=/<= without the if statement. I was trying to find a way to avoid calling `reduce` in such a similar way twice, but couldn't come up with anything because of the scope of `reduce`.

Reply:
Sorry, i meant for `opponent` not `best_move`